src/components/feedback/Modal.tsx

```diff
@@ -0,0 +1,136 @@
+import classnames from 'classnames';
+import type { RefObject } from 'preact';
+import { useEffect, useLayoutEffect } from 'preact/hooks';
+
+import { useElementShouldClose } from '../../hooks/use-element-should-close';
+import { useSyncedRef } from '../../hooks/use-synced-ref';
+import { useUniqueId } from '../../hooks/use-unique-id';
+
+import type { PresentationalProps } from '../../types';
+import { downcastRef } from '../../util/typing';
+
+import Overlay from '../layout/Overlay';
+import Panel from '../layout/Panel';
+import type { PanelProps } from '../layout/Panel';
+
+type ComponentProps = {
+  /**
+   * Modal _should_ be provided with a close handler. We have a few edge use
+   * cases, however, in which we need to render a "non-closeable" modal.
+   */
+  onClose?: () => void;
+  width?: 'sm' | 'md' | 'lg' | 'custom';
+  /**
+   * Element that should take focus when the Dialog is first rendered. When not
+   * provided, the dialog's outer element will take focus. Setting this prop to
+   * `null` will opt out of focus routing.
+   */
+  initialFocus?: RefObject<HTMLElement | null> | null;
+};
+
+export type ModalProps = PresentationalProps & ComponentProps & PanelProps;
+
+const noop = () => {};
+
+/**
+ * Show a modal
+ */
+const ModalNext = function Modal({
+  children,
+  width = 'md',
+
+  classes,
+  elementRef,
+  initialFocus,
+
+  // Forwarded to Panel
+  buttons,
+  icon,
+  onClose,
+  title,
+
+  ...htmlAttributes
+}: ModalProps) {
+  const modalRef = useSyncedRef(elementRef);
+  const closeHandler = onClose ?? noop;
+  useElementShouldClose(modalRef, true, closeHandler);
+
+  const dialogDescriptionId = useUniqueId('dialog-description');
+
+  useEffect(() => {
+    // Route focus unless `initialFocus` is set to `null`
+    if (initialFocus !== null) {
+      const focusEl = initialFocus?.current as HTMLInputElement;
+      if (focusEl && !focusEl.disabled) {
+        focusEl.focus();
+      } else {
+        // The `initialFocus` prop has not been set, so use automatic focus
+        // handling. Modern accessibility guidance is to focus the dialog itself
+        // rather than trying to be smart about focusing a particular control
+        // within the dialog.
+        modalRef.current?.focus();
+      }
+    }
```

Comment:
Can I suggest not nesting `if`s? I find it easier to use early returns so that you can mentally discard conditions as soon as possible, as you "follow" the code. When nesting, you need to "drag" all of them.

```ts
if (initialFocus === null) {
  return;
}

const focusEl = initialFocus?.current as HTMLInputElement;
if (focusEl && !focusEl.disabled) {
    focusEl.focus();
} else {
  // ...
  modalRef.current?.focus();
}
```

Reply:
Sure! I kind of blindly cut-and-pasted this from a previous implementation so it's good to get fresh eyes on it. I agree. Guards are easier to reason about.